File: src/plnflr/engine/instructions.py
"""Polish row-by-row install copy."""

from __future__ import annotations

from collections import defaultdict
from dataclasses import replace

from plnflr.domain.models import LayoutPlan, Piece


def instructions_for(pieces: tuple[Piece, ...], *, gap_mm: int, has_holes: bool) -> tuple[str, ...]:
    by_row: dict[int, list[Piece]] = defaultdict(list)
    for piece in pieces:
        by_row[piece.row_index].append(piece)
    lines = [f"Dylatacja {gap_mm} mm wokół obrysu" + (" i przeszkód." if has_holes else ".")]
    for row in sorted(by_row):
        row_pieces = sorted(by_row[row], key=lambda p: p.install_order)
        width = row_pieces[0].width_mm
        clips = sum(1 for p in row_pieces if p.kind in {"clip", "rip"})
        start = row_pieces[0].length_mm
        end = row_pieces[-1].length_mm
        note = ""
        if clips:
            note = " Przytnij do ściany / otworu."
        lines.append(
            f"Rząd {row + 1} (szer. {width} mm): start {start} mm → "
            f"{len(row_pieces)} szt. → koniec {end} mm.{note}"
        )
    return tuple(lines)


def attach(plan: LayoutPlan) -> LayoutPlan:
    return replace(
        plan,
        rows_instruction_pl=instructions_for(
            plan.pieces, gap_mm=plan.gap_mm, has_holes=bool(plan.room.holes)
        ),
    )

File: Kernel/Sources/PlnFlrLayout/Instructions.swift
func instructionsFor(_ pieces: [Piece], gapMm: Int, hasHoles: Bool) -> [String] {
    var byRow: [Int: [Piece]] = [:]
    for piece in pieces {
        byRow[piece.rowIndex, default: []].append(piece)
    }
    var lines = [
        "Dylatacja \(gapMm) mm wokół obrysu" + (hasHoles ? " i przeszkód." : ".")
    ]
    for row in byRow.keys.sorted() {
        let rowPieces = byRow[row]!.sorted { $0.installOrder < $1.installOrder }
        let width = rowPieces[0].widthMm
        let clips = rowPieces.filter { $0.kind == .clip || $0.kind == .rip }.count
        let start = rowPieces[0].lengthMm
        let end = rowPieces[rowPieces.count - 1].lengthMm
        var note = ""
        if clips > 0 { note = " Przytnij do ściany / otworu." }
        lines.append(
            "Rząd \(row + 1) (szer. \(width) mm): start \(start) mm → \(rowPieces.count) szt. → koniec \(end) mm.\(note)"
        )
    }
    return lines
}
